TACC/Core-CMS
# GH-89: (Minimal) System Monitor Plugin (a.k.a. SysMon)

# Important

The PR https://github.com/TACC/Core-CMS-Resources/pull/67 is dependent on this PR.

# To Do

- [x] 0d9491f (Optional) Rename `sysmon` to `system_monitor`.\*
- [x] a36e978 [Remove unused `get_direction_classname` import.](https://github.com/TACC/Core-CMS/pull/297#discussion_r687169681)
- [x] (no change) [Revisit `setStatusStyle()`.](https://github.com/TACC/Core-CMS/pull/297#discussion_r687187046)
- [x] (no change) [Re-evaluate where to place constants.](https://github.com/TACC/Core-CMS/pull/297#discussion_r687176885)
- [x] 5ed2892, a2807a2 [Resolve unexpected plugin installs (and search for related problems)](https://github.com/TACC/Core-CMS/pull/297#pullrequestreview-731178122)
- [x] 7220bc1 [Ensure multiple instances (a) can exist or (b) are not permitted to exist.†](https://github.com/TACC/Core-CMS/pull/297#pullrequestreview-731178122)
- [x] 333918d Fix warning badge missing background color.

<details>
<summary>Footnotes</summary>

\* Rename to be consistent with [upcoming `system_specs`](https://github.com/TACC/Core-CMS/compare/task/GH-88-system-specs) and possible `system_...`.
† Either [(a) do not use `GetElementById`](https://github.com/TACC/Core-CMS/commit/7220bc1fbe038973bdd85729beea871f17385029) or ~~(b) limit plugin to one instance~~. _Try not to over-engineer the JavaScript, because #295 will delete it._

</details>

# Overview

Implement _minimal_ SysMon plugin using polished CSS and JavaScript.

# Issues

* #89

# Changes

- Create plugin from snippet.

# Screenshots

## Core

<details open>
<summary>Toggle visibility</summary>

### Form
![Sysmon Plugin - Admin Form](https://user-images.githubusercontent.com/62723358/127896948-9eea5379-cbe8-480a-9d59-fd7716c1a25e.png)

### Render
![Sysmon Plugin - Core Styles](https://user-images.githubusercontent.com/62723358/127896950-2ac2a6c4-0528-4939-8468-dc18951b2a16.png)

### Structure
![System Plugin - in Structure](https://user-images.githubusercontent.com/62723358/127896952-cd6f7b23-16e0-4dd1-aafb-6da6f9c600a0.png)

</details>

## Frontera

<details>
<summary>Toggle visibility</summary>

### Render
![Sysmon Plugin - Frontera Home - Render](https://user-images.githubusercontent.com/62723358/127908190-224ee788-0719-4627-9f84-0b0d1356fb95.png)

### Structure
![Sysmon Plugin - Frontera Home - Structure](https://user-images.githubusercontent.com/62723358/127908202-760326b5-7e49-4cbe-b2b1-067305688fa1.png)

</details>

# Testing

> Last recorded deploy to [Frontera (dev)](https://dev.fronteraweb.tacc.utexas.edu/) was [Build #383 (Aug 2, 2021 1:58:12 PM)](https://jenkins01.tacc.utexas.edu/job/Core_Portal_Deploy/383/).

> __Important__: If you tested this plugin before it was renamed from "Sysmon" to "System Monitor", then you must follow these cleanup steps to prevent a CMS server crash:
>
> <details>
> <summary><strong>Steps for Plugin Rename</strong></summary>
>
> 1. Checkout commit or branch that does not crash server.
>     - If you have only been testing this plugin, then a36e978 should work.
>     - If you have been testing several plugins, then `site/frontera/main` might work.
> 2. Delete "Sysmon" plugin instances:
>     ```
>     python3 manage.py cms uninstall plugins TaccsiteSysmonPlugin
>     ```
> 3. Checkout `task/GH-89` branch at latest commit.
> 4. Perform database migrations for new "System Monitor" plugin:
>     ```
>     python3 manage.py migrate
>     ```
> 5. Re-add deleted plugin instances (as necessary for testing) using new "System Monitor" plugin.
> 6. Rebuild assets and collect static files:
>     ```
>     npm run build && docker exec -it core_cms python manage.py collectstatic --no-input
>     ```
>
> </details>

1. If no environment, then [prepare to test CMS changes](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes).
2. In CMS container, run `python3 manage.py migrate`.
3. On CMS website, logged in as admin, add "TACC Site" > "System Monitor" plugin.
4. Reload page (to trigger client-side\* data retrieval).
5. Ensure the following:
    - styles match system monitor on [Old Homepage][old-home] / [Phase 0 Homepage][new-home]
        - the text may be larger; this is acceptable†
    - data shown is for system chosen
    - plugin description (next to name) in Structure view is the system chosen
6. Add "Generic" > "Text" plugin.
7. In Text plugin, in "CMS Plugins" dropdown, confirm "TACC Site" > "System Monitor" plugin is **not** available.
8. Add another "TACC Site" > "System Monitor" plugin, and choose "Stampede2".
9. Ensure Stampede2 monitor—
    - if on local server—has "Warning" state
    - if on remote server—is populated with live data

<details>
<summary>Footnotes</summary>

\* I hope to implement server-side retrieval later via #295.
† Different projects and templates have different root font size. ([Values have been standardized](https://confluence.tacc.utexas.edu/x/nh4FDg), but implementation has not caught up to standards.)

</details>

# Notes

This is a _minimal_ version of a functionally better plugin, [bitbucket:rochaa/tacc-sysmon][bb-sysmon]. @tacc-wbomar has deadline to finish #88 (which requires #89) this week. So #295 has been created to enhance this plugin later.

[bb-sysmon]: https://bitbucket.org/rochaa/tacc-sysmon
[old-home]: https://frontera-portal.tacc.utexas.edu/home-before-2021-07/
[new-home]: https://frontera-portal.tacc.utexas.edu/
